tempesta_fw/sock.c

```diff
@@ -194,6 +194,9 @@ ss_do_send(struct sock *sk, SsSkbList *skb_list)
 	       smp_processor_id(), __func__,
 	       sk, tcp_send_head(sk), sk->sk_state);
 
+	if ((flags & SS_SEND_F_CONN_CLOSE) && tcp_close_state(sk))
+		ss_tcp_shutdown(sk);
```

Comment:
Actually `ss_tcp_shutdown()` is already partially [implements `ss_do_close()`](https://github.com/tempesta-tech/tempesta/blob/master/tempesta_fw/sock.c#L339).

It seems we should at least add `tcp_close_state()` call to `ss_tcp_shutdown()` since we have to move the socket to FIN-WAIT-1 state when we send FIN, but it seems `tcp_push()` doesn't analyze the flag for transmitted skbs. I think TCP stack won't close the connection properly when it receives ACK for the FIN. Probably we also have to do `sk->sk_shutdown = SHUTDOWN_MASK`.

Since we don't want and can't to send more data on the socket, we can do `sk_mem_reclaim(sk)` as well.